hw/ip/otbn/util/static_check.py

```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import re
+import struct
+import sys
+from typing import Dict, Iterator, List, Optional, Tuple
+
+from shared.elf import read_elf
+from shared.insn_yaml import Insn, load_insns_yaml
+
+# Load the insns.yml file at module load time.
+try:
+    INSNS_FILE = load_insns_yaml()
+except RuntimeError as err:
+    sys.stderr.write('{}\n'.format(err))
+    sys.exit(1)
+
+
+class OTBNProgram:
+    def __init__(self, symbols: Dict[str, int], insns: Dict[int, int],
+                 data: Dict[int, int]):
+        self.symbols = symbols  # label -> PC
+        self.insns = insns  # PC -> opcode
+        self.data = data  # addr -> data (32b word)
+
+    def _get_pc_at_symbol(self, symbol: str) -> int:
+        return self.symbols[symbol]
+
+    def _get_opcode_at_pc(self, pc: int) -> int:
+        return self.insns[pc]
+
+    def _get_data_at_addr(self, addr: int) -> int:
+        return self.data[addr]
+
+    def _get_insn(self, pc: int) -> Insn:
+        opcode = self._get_opcode_at_pc(pc)
+        mnem = INSNS_FILE.mnem_for_word(opcode)
+        if mnem is None:
+            return IllegalInsn(pc, word, 'No legal decoding')
+
+        return INSNS_FILE.mnemonic_to_insn.get(mnem)
+
+    def _get_operands(self, pc: int) -> Dict[str, int]:
+        insn = self._get_insn(pc)
+        opcode = self._get_opcode_at_pc(pc)
+        enc_vals = insn.encoding.extract_operands(opcode)
+        return insn.enc_vals_to_op_vals(pc, enc_vals)
+
+
+def _decode_mem(base_addr: int, data: bytes) -> Dict[int, int]:
+    '''Interpret memory bytes as a dictionary indexed by address/PC'''
+    assert len(data) & 3 == 0
+    return {(base_addr + 4 * offset): int_val[0]
+            for offset, int_val in enumerate(struct.iter_unpack('<I', data))}
+
+
+def decode_elf(path: str) -> OTBNProgram:
+    '''Read ELF file at path and decode contents into an OTBNProgram instance
+
+    Returns the OTBNProgram instance representing the program in the ELF file.
+    '''
+    (imem_bytes, dmem_bytes, symbols) = read_elf(path)
+
+    insns = _decode_mem(0, imem_bytes)
+    data = _decode_mem(0, dmem_bytes)
+
+    return OTBNProgram(symbols, insns, data)
+
+
+def _check_call_stack_insn(insn, operands):
+    '''Returns true iff insn with these operands is safe from CALL_STACK errors.
+
+    Analyzes the possible errors from insn.errs and returns False if one of the
+    error conditions could potentially be triggered. Note that many of the
+    errors only happen when the call stack is in a certain state (e.g. full or
+    empty); this check does not attempt to track call stack use, and will
+    return False if there exists *any* call stack that could trigger the error.
+    '''
+    if insn.errs is None:
+        return True
+    for err in insn.errs:
```

Comment:
I think there's probably an easier way to do this than string matching. The operands in `insn.operands` have types, so I think you just want to check that no GPR register operand is equal to 1 unless this is a JAL or JALR.

Reply:
Now updated to do that, which simplifies the code considerably; I wasn't sure if it was okay to error on _any_ use of x1 outside of JAL/JALR, or only ones that could cause CALL_STACK errors. But it's probably true that you really shouldn't be using x1 for anything else, lest your code have other issues.

Reply:
Assuming I got the documentation right, anything that can read or write `x1` should say that it can cause a call stack error (either from under- or over-flow). Please let me know if I missed something! :-)